### PR TITLE
[REVIEW] Python bindings for rolling windows

### PR DESCRIPTION
**Pull request by shwina**
Rolling window operations

- [x] Simple numerical rolling windows for series
- [x] Support and test all libcudf agg types
- [x] Add DataFrame rolling
- [x] Support for min_periods keyword
- [x] Support for center=True/False
- [x] Datetime support*
- [ ] Support the win_type argument and set default to `boxcar`
- [x] Docstring with examples

* refers to Pandas feature that allows `window` to be an offset like `'2s'` (only valid for datetime indexes).

Not supported by Pandas:
- [ ] Support for asymmetric windows
- [ ] Support for dynamic rolling windows

Others:
- [ ] Add `Groupby.rolling()`

Keep rollin' rollin' rollin' rollin'

### CODE REVIEW

File: python/cudf/bindings/rolling.pyx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+from cudf.dataframe.column import Column
+
+from cudf.bindings.cudf_cpp cimport *
+from cudf.bindings.cudf_cpp import *
+from cudf.bindings.rolling cimport *
+
+
+def apply_rolling(inp, window, min_periods, center, op):
+
+    cdef gdf_column *inp_col
```
[the remark this answers is not in the document]
Comment (shwina) [reply python/cudf/bindings/rolling.pyx]: free'd below

File: python/cudf/utils/cudautils.py [hunk 2]
```diff
@@ -964,3 +964,21 @@ def boolean_array_to_index_array(bool_array):
...
+        while j > -1:
+            if (arr[i] - arr[j]) >= offset:
+                break
+            j -= 1
+        window_sizes[i] = i - j
+
+
+def window_sizes_from_offset(arr, offset):
+    window_sizes = rmm.device_array(shape=(arr.shape), dtype="int32")
+    gpu_window_sizes_from_offset.forall(arr.size)(arr, window_sizes, offset)
```
Comment (kkraus14) [inline python/cudf/utils/cudautils.py]: You should guard here against a 0 sized kernel launch:
```suggestion
    if arr.size > 0:
        gpu_window_sizes_from_offset.forall(arr.size)(arr, window_sizes, offset)
```

**Comment by shwina** [comment]
Hmm, it looks like a Pandas bug in the CI version of Pandas is causing tests to fail. I'm working locally with Pandas 0.24.2, but I think CI is working with 0.23.4:

**Pandas 0.23.4**

```
In [1]: import numpy as np
In [2]: import pandas as pd
In [3]: psr = pd.Series( 
   ...:     [1, 2, 4, 4, np.nan, 9], 
   ...:     index=[ 
   ...:         pd.Timestamp('20190101 09:00:00'), 
   ...:         pd.Timestamp('20190101 09:00:01'), 
   ...:         pd.Timestamp('20190101 09:00:02'), 
   ...:         pd.Timestamp('20190101 09:00:04'), 
   ...:         pd.Timestamp('20190101 09:00:07'), 
   ...:         pd.Timestamp('20190101 09:00:08') 
   ...:     ] 
   ...: ) 
   ...: 
In [4]: psr.rolling('2s').max()                                                                                                                                                                                 
Out[4]: 
2019-01-01 09:00:00    1.0
2019-01-01 09:00:01    2.0
2019-01-01 09:00:02    4.0
2019-01-01 09:00:04    4.0
2019-01-01 09:00:07    NaN
2019-01-01 09:00:08    NaN
dtype: float64
```

**Pandas 0.24.2**

```
In [1]: import numpy as np
In [2]: import pandas as pd
In [3]: psr = pd.Series( 
   ...:     [1, 2, 4, 4, np.nan, 9], 
   ...:     index=[ 
   ...:         pd.Timestamp('20190101 09:00:00'), 
   ...:         pd.Timestamp('20190101 09:00:01'), 
   ...:         pd.Timestamp('20190101 09:00:02'), 
   ...:         pd.Timestamp('20190101 09:00:04'), 
   ...:         pd.Timestamp('20190101 09:00:07'), 
   ...:         pd.Timestamp('20190101 09:00:08') 
   ...:     ] 
   ...: )
In [4]: psr.rolling('2s').max()                                                                                                                                                                                 
Out[4]: 
2019-01-01 09:00:00    1.0
2019-01-01 09:00:01    2.0
2019-01-01 09:00:02    4.0
2019-01-01 09:00:04    4.0
2019-01-01 09:00:07    NaN
2019-01-01 09:00:08    9.0
dtype: float64
```

@kkraus14 Can we mark the test xfail for now?

**Comment by kkraus14** [comment]
Yea, let's just xfail it for now.

**Pull request merged**